src/Hpack/Exception.hs

```diff
@@ -0,0 +1,71 @@
+{-# LANGUAGE RecordWildCards #-}
+
+-- | Exceptions thrown by the Hpack library.
+module Hpack.Exception
+  ( HpackException (..)
+    -- * Re-export of types used in Hpack exceptions
+  , FilePath
+  , ParseException (..)
+  , ProgramName (..)
+  , Status (..)
+  , Version (..)
+  , URL
+  , YamlException (..)
+  , YamlMark (..)
+  ) where
+
+import           Control.Exception (Exception (..))
+import qualified Data.ByteString.Char8 as B
+import           Data.List (intercalate)
+import           Data.Typeable (Typeable)
+import           Data.Version (Version (..), showVersion)
+import           Data.Yaml
+                   (ParseException (..), YamlException (..), YamlMark (..))
+import           Network.HTTP.Types.Status (Status (..))
+
+import           Hpack.ProgramName (ProgramName (..))
+
+-- | Type synonyn representing URLs.
+type URL = String
+
+-- | Type representing exceptions thrown by functions exported by the modules of
+-- the Hpack library.
+data HpackException
+  = CycleInDefaultsException ![FilePath]
+  | DecodeValueException !FilePath !String
+  | DefaultsFileNotFound !FilePath
+  | DefaultsFileUrlNotFound !URL
+  | DownloadingFileFailed !URL !Status
+  | HpackParseException !FilePath !ParseException
```

Comment:
I'll address this in a comment below, but the data constructors of type `HpackError` (as it now is) now only use values of types that are instances of `Eq`, so that `HpackError` can be an instance of `Eq`.